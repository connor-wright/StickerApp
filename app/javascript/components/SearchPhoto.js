import React from "react";
class SearchPhoto extends React.Component {
  render () {
    const {url, onClick, xi, active} = this.props;
    const style = active? {border: '2px solid blue'} : {};
    return (
      <img style={style} src={url} onClick={() => onClick(xi)}/>
    );
  }
}

export default SearchPhoto;
