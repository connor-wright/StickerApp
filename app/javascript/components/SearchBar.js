import React from "react";
import "./APIInterface";
import SearchPhoto from "./SearchPhoto";
import {SearchImgs} from "./APIInterface"

class SearchBar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {value: '', stickers: []};
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.onClick      = this.onClick.bind(this);
  }
  
  onClick(xi) {
    const images = this.state.stickers;
    this.props.setActiveId(images[xi].id);
    this.setState({stickers: 
      images.map((image, index) => 
        ({
          active: index == xi? true: false, id: image.id, url: image.url
        })
      )
    });
  }
  
  handleChange(event) {
    this.setState({value: event.target.value});
  }
  
  handleSubmit(event) {
    if(this.state.value)
    {
      let AddStickers = (photos) => this.setState(() => ({
        stickers: photos.map(photo => 
        {
          return {active: false, id: photo.id, url: photo.src.small};
        })
      }));
      SearchImgs(this.state.value).then((result) => AddStickers(result.photos));
    }
    event.preventDefault();
  }
  
  render () {
    const {stickers} = this.state;
    return (
      <div className='SearchBar'>
        <form onSubmit={this.handleSubmit}>
          <label>
            search:
            <input type="text" data-testid='searchInput' value={this.state.value} onChange={this.handleChange} />
          </label>
            <input type="submit" value="Submit" />
          </form>
          {stickers.map((image, xi) =>
            <SearchPhoto 
              data-testid='searchPhoto'
              setActiveId={this.props.setActiveId}
              url={image.url}
              xi={xi}
              key={xi}
              active={image.active}
              onClick={this.onClick}/>
          )}
      </div>
    );
  }
}

export default SearchBar;