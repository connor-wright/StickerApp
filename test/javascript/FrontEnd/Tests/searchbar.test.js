import React from 'react';
import SearchBar from 'SearchBar';
import SearchPhoto from 'SearchPhoto';
import { mount } from 'enzyme';
import {runAllPromises} from '../Utils/utils';
const BackendAPI = require('BackendAPI');

let activeId;
let setActiveId;
let searchBar;

jest.mock('BackendAPI');
const SearchImgs = BackendAPI.SearchImgs;

describe('<SearchBar />', () => {
  
  beforeEach(() =>{
    activeId = '';
    setActiveId = newActiveId => activeId = newActiveId;
    searchBar = mount(<SearchBar setActiveId={setActiveId}/>);
  });
  
  afterEach(() => {
    searchBar.unmount();
  });
  
  it('should search triggers photo search to backend', async () =>{
    //arrange
    SearchImgs.mockResolvedValue({photos: []});
    
    //act
    const searchForm = searchBar.find('.searchInput');
    await searchForm.simulate('change', {target: {value: 'whatever'}});
    await searchForm.simulate('submit');
    expect(SearchImgs).toHaveBeenCalled();
  });
  
  it('should create search images after submit', async ()=> {
    //arrange
    SearchImgs.mockResolvedValue({photos: [
      {id: 'Taco', src: {small: 'https://imgur.com/gallery/3nSMulw'}}
    ]});
    
    //act
    const searchForm = searchBar.find('.searchInput');
    await searchForm.simulate('change', {target: {value: 'whatever'}});
    await searchForm.simulate('submit');
    
    //assert
    await runAllPromises();
    searchBar.update();
    expect(searchBar).toMatchSnapshot();
  });
  
  it('should pass the query to the api after submit', async () => {
    //arrange
    const expectedQuery = 'tacosAreUnique';
    
    //act
    const searchForm = searchBar.find('.searchInput');
    await searchForm.simulate('change', {target: {value: expectedQuery}});
    await searchForm.simulate('submit');
    
    //assert
    expect(SearchImgs.mock.calls).toContainEqual([expectedQuery]);
  });
  
  it('should set the active ID after clicking on an image', async () => {
    //arrange
    const expectedId = 'TacosAreTheBest';
    SearchImgs.mockResolvedValue({photos: [
      {id: expectedId, src: {small: 'https://imgur.com/gallery/3nSMulw'}}
    ]});
    const searchForm = searchBar.find('.searchInput');
    await searchForm.simulate('change', {target: {value: 'whatever'}});
    await searchForm.simulate('submit');
    await runAllPromises();
    searchBar.update();
    
    //act
    searchBar.find(SearchPhoto).simulate('click');
    
    //assert
    expect(expectedId).toEqual(activeId)
  });
});