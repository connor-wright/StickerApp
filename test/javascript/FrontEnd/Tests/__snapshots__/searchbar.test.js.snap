// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> should create search images after submit 1`] = `
<SearchBar
  setActiveId={[Function]}
>
  <div
    className="SearchBar"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        search:
        <input
          className="searchInput"
          onChange={[Function]}
          type="text"
          value="whatever"
        />
      </label>
      <input
        type="submit"
        value="Submit"
      />
    </form>
    <SearchPhoto
      active={false}
      key="0"
      onClick={[Function]}
      url="https://imgur.com/gallery/3nSMulw"
      xi={0}
    >
      <img
        onClick={[Function]}
        src="https://imgur.com/gallery/3nSMulw"
        style={Object {}}
      />
    </SearchPhoto>
  </div>
</SearchBar>
`;
